// @tencent/winston-taf/transport/util/date-format.js
var assert = require('assert');

var Base = function(logPattern, name, interval, timePattern) {
	if (!(this instanceof Base)) {
		return new Base(interval, logPattern, timePattern);
	}

	this.logPattern = logPattern;
	this.name = name;

	if (interval && timePattern) { 
		this.interval = parseInt(interval);
		assert(this.interval > 0, 'interval must be large than 0');

		this.timePattern = timePattern;
	}
};

var LogByDay = function(interval, pattern) {
	return new Base(pattern || '%Y%m%d', 'day', interval || 1, '%Y%m%d');
};

var LogByHour = function(interval, pattern) {
	assert(!interval || (interval < 24 && interval > 0), 'interval can\'t be large than 24, less than 0');
	
	return new Base(pattern || '%Y%m%d%H', 'hour', interval || 1, '%Y%m%d%H');
};

var LogByMinute = function(interval, pattern) {
	assert(!interval || interval < 60 && interval > 0, 'interval can\'t be large than 60, less than 0');
	return new Base(pattern || '%Y%m%d%H%M', 'minute', interval || 10, '%Y%m%d%H%M');
};

var LogByCustom = function(pattern) {
	assert(pattern, 'pattern not defined');
	return new Base(pattern, 'custom');
};

exports.LogByDay = LogByDay;
exports.LogByHour = LogByHour;
exports.LogByMinute = LogByMinute;
exports.LogByCustom = LogByCustom;
