// @tencent/taf-rpc-jce/RequestFJce.js
// **********************************************************************
// Parsed By TAFParser(3.2.2.11), Generated By Jce2Node(20180308)
// TAFParser Maintained By <WSRD> and Jce2Node Maintained By <superzheng>
// Generated from "RequestF.jce" by Structure Mode
// **********************************************************************

"use strict";

var assert    = require("assert");
var TafStream = require("@tencent/taf-stream");

var taf = taf || {};
module.exports.taf = taf;

taf.RequestPacket = function() {
    this.iVersion = 0;
    this.cPacketType = 0;
    this.iMessageType = 0;
    this.iRequestId = 0;
    this.sServantName = "";
    this.sFuncName = "";
    this.sBuffer = new TafStream.BinBuffer;
    this.iTimeout = 0;
    this.context = new TafStream.Map(TafStream.String, TafStream.String);
    this.status = new TafStream.Map(TafStream.String, TafStream.String);
    this._classname = "taf.RequestPacket";
};
taf.RequestPacket._classname = "taf.RequestPacket";
taf.RequestPacket._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.RequestPacket._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.RequestPacket._readFrom = function (is) {
    var tmp = new taf.RequestPacket();
    tmp.iVersion = is.readInt16(1, true, 0);
    tmp.cPacketType = is.readInt8(2, true, 0);
    tmp.iMessageType = is.readInt32(3, true, 0);
    tmp.iRequestId = is.readInt32(4, true, 0);
    tmp.sServantName = is.readString(5, true, "");
    tmp.sFuncName = is.readString(6, true, "");
    tmp.sBuffer = is.readBytes(7, true, TafStream.BinBuffer);
    tmp.iTimeout = is.readInt32(8, true, 0);
    tmp.context = is.readMap(9, true, TafStream.Map(TafStream.String, TafStream.String));
    tmp.status = is.readMap(10, true, TafStream.Map(TafStream.String, TafStream.String));
    return tmp;
};
taf.RequestPacket.prototype._writeTo = function (os) {
    os.writeInt16(1, this.iVersion);
    os.writeInt8(2, this.cPacketType);
    os.writeInt32(3, this.iMessageType);
    os.writeInt32(4, this.iRequestId);
    os.writeString(5, this.sServantName);
    os.writeString(6, this.sFuncName);
    os.writeBytes(7, this.sBuffer);
    os.writeInt32(8, this.iTimeout);
    os.writeMap(9, this.context);
    os.writeMap(10, this.status);
};
taf.RequestPacket.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.RequestPacket.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.RequestPacket.prototype.toObject = function() { 
    return {
        "iVersion" : this.iVersion,
        "cPacketType" : this.cPacketType,
        "iMessageType" : this.iMessageType,
        "iRequestId" : this.iRequestId,
        "sServantName" : this.sServantName,
        "sFuncName" : this.sFuncName,
        "sBuffer" : this.sBuffer.toObject(),
        "iTimeout" : this.iTimeout,
        "context" : this.context.toObject(),
        "status" : this.status.toObject()
    };
};
taf.RequestPacket.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("iVersion") && (this.iVersion = json.iVersion);
    json.hasOwnProperty("cPacketType") && (this.cPacketType = json.cPacketType);
    json.hasOwnProperty("iMessageType") && (this.iMessageType = json.iMessageType);
    json.hasOwnProperty("iRequestId") && (this.iRequestId = json.iRequestId);
    json.hasOwnProperty("sServantName") && (this.sServantName = json.sServantName);
    json.hasOwnProperty("sFuncName") && (this.sFuncName = json.sFuncName);
    json.hasOwnProperty("sBuffer") && (this.sBuffer.readFromObject(json.sBuffer));
    json.hasOwnProperty("iTimeout") && (this.iTimeout = json.iTimeout);
    json.hasOwnProperty("context") && (this.context.readFromObject(json.context));
    json.hasOwnProperty("status") && (this.status.readFromObject(json.status));
};
taf.RequestPacket.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.RequestPacket.new = function () {
    return new taf.RequestPacket();
};
taf.RequestPacket.create = function (is) {
    return taf.RequestPacket._readFrom(is);
};

taf.ResponsePacket = function() {
    this.iVersion = 0;
    this.cPacketType = 0;
    this.iRequestId = 0;
    this.iMessageType = 0;
    this.iRet = 0;
    this.sBuffer = new TafStream.BinBuffer;
    this.status = new TafStream.Map(TafStream.String, TafStream.String);
    this.sResultDesc = "";
    this.context = new TafStream.Map(TafStream.String, TafStream.String);
    this._classname = "taf.ResponsePacket";
};
taf.ResponsePacket._classname = "taf.ResponsePacket";
taf.ResponsePacket._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.ResponsePacket._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.ResponsePacket._readFrom = function (is) {
    var tmp = new taf.ResponsePacket();
    tmp.iVersion = is.readInt16(1, true, 0);
    tmp.cPacketType = is.readInt8(2, true, 0);
    tmp.iRequestId = is.readInt32(3, true, 0);
    tmp.iMessageType = is.readInt32(4, true, 0);
    tmp.iRet = is.readInt32(5, true, 0);
    tmp.sBuffer = is.readBytes(6, true, TafStream.BinBuffer);
    tmp.status = is.readMap(7, true, TafStream.Map(TafStream.String, TafStream.String));
    tmp.sResultDesc = is.readString(8, false, "");
    tmp.context = is.readMap(9, false, TafStream.Map(TafStream.String, TafStream.String));
    return tmp;
};
taf.ResponsePacket.prototype._writeTo = function (os) {
    os.writeInt16(1, this.iVersion);
    os.writeInt8(2, this.cPacketType);
    os.writeInt32(3, this.iRequestId);
    os.writeInt32(4, this.iMessageType);
    os.writeInt32(5, this.iRet);
    os.writeBytes(6, this.sBuffer);
    os.writeMap(7, this.status);
    os.writeString(8, this.sResultDesc);
    os.writeMap(9, this.context);
};
taf.ResponsePacket.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.ResponsePacket.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.ResponsePacket.prototype.toObject = function() { 
    return {
        "iVersion" : this.iVersion,
        "cPacketType" : this.cPacketType,
        "iRequestId" : this.iRequestId,
        "iMessageType" : this.iMessageType,
        "iRet" : this.iRet,
        "sBuffer" : this.sBuffer.toObject(),
        "status" : this.status.toObject(),
        "sResultDesc" : this.sResultDesc,
        "context" : this.context.toObject()
    };
};
taf.ResponsePacket.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("iVersion") && (this.iVersion = json.iVersion);
    json.hasOwnProperty("cPacketType") && (this.cPacketType = json.cPacketType);
    json.hasOwnProperty("iRequestId") && (this.iRequestId = json.iRequestId);
    json.hasOwnProperty("iMessageType") && (this.iMessageType = json.iMessageType);
    json.hasOwnProperty("iRet") && (this.iRet = json.iRet);
    json.hasOwnProperty("sBuffer") && (this.sBuffer.readFromObject(json.sBuffer));
    json.hasOwnProperty("status") && (this.status.readFromObject(json.status));
    json.hasOwnProperty("sResultDesc") && (this.sResultDesc = json.sResultDesc);
    json.hasOwnProperty("context") && (this.context.readFromObject(json.context));
};
taf.ResponsePacket.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.ResponsePacket.new = function () {
    return new taf.ResponsePacket();
};
taf.ResponsePacket.create = function (is) {
    return taf.ResponsePacket._readFrom(is);
};
