// @tencent/taf-rpc/core/server/BindAdapter.js
var assert         = require("assert");
var Endpoint       = require("@tencent/taf-utils").Endpoint;
var TimeProvider   = require("@tencent/taf-utils").timeProvider;
var TafError       = require("../util/TafError.js").TafError;
var NetThread      = require("./NetThread.js").NetThread;
var JceCurrent     = require("./JceCurrent.js").JceCurrent;

///////////////////////////////////////////////监听端口功能包裹类///////////////////////////////////////////////////////
var BindAdapter = function () {
    this._objname       = "";           //BindAdapter的名字
    this._servant       = "";           //当前端口上的servant名字
    this._endpoint      = undefined;    //当前端口监听地址

    this._protocolName  = "";
    this._protocol      = undefined;    //当前端口上的请求包协议解析器
    this._maxconns      = 1024;         //当前端口上最大的连接数

    this._netthread     = undefined;    //当前端口上的监听服务
    this._handleImp     = undefined;    //当前端口上的逻辑处理对象实例

    this._authKeyMap    = {};           //存储此adapter上的appkey、secretkey(两次sha1后) 的map
    this._authFn        = undefined; //自定义鉴权函数(解密串)
};
module.exports.BindAdapter = BindAdapter;

//创建对外属性设置接口
BindAdapter.prototype.__defineGetter__("name", function () { return this._objname; });
BindAdapter.prototype.__defineSetter__("name", function (value) { this._objname = value; });

BindAdapter.prototype.__defineGetter__("authKeyMap", function () { return this._authKeyMap; });
BindAdapter.prototype.__defineSetter__("authKeyMap", function (value) { this._authKeyMap = value; });

BindAdapter.prototype.__defineGetter__("servantName", function () { return this._servant; });
BindAdapter.prototype.__defineSetter__("servantName", function (value) { this._servant = value; });

BindAdapter.prototype.__defineGetter__("maxconns", function () { return this._maxconns; });
BindAdapter.prototype.__defineSetter__("maxconns", function (value) { this._maxconns = value; });

BindAdapter.prototype.__defineGetter__("endpoint", function () { return this._endpoint; });
BindAdapter.prototype.__defineSetter__("endpoint", function (value) { this._endpoint = value; });

BindAdapter.prototype.__defineGetter__("protocolName", function () { return this._protocolName; });
BindAdapter.prototype.__defineSetter__("protocolName", function (value) { this._protocolName = value; });

BindAdapter.prototype.__defineGetter__("protocol", function () { return this._protocol; });
BindAdapter.prototype.__defineSetter__("protocol", function (value) { this._protocol = value; });

BindAdapter.prototype.__defineGetter__("handleImp", function () { return this._handleImp; });
BindAdapter.prototype.__defineSetter__("handleImp", function (value) { this._handleImp = value; });

BindAdapter.prototype.isTafProtocol = function () { //判断是否是TAF协议
    return this._protocolName === "taf";
};

/////鉴权相关
BindAdapter.prototype.addAuthKeyMap = function (sAccessKey, sHashSecretKey2) {
    this._authKeyMap[sAccessKey] = sHashSecretKey2;
}

BindAdapter.prototype.updateAuthKeyMap = function (authKeyMap) {
    this._authKeyMap = authKeyMap;
}

BindAdapter.prototype.getSk = function (sAccessKey) {
    return this._authKeyMap[sAccessKey];
}

BindAdapter.prototype.setAuthFn = function(fn){
    this._authFn = fn;
}

BindAdapter.prototype.start = function () {
    assert(this._endpoint.sProtocol === "tcp" || this._endpoint.sProtocol === "udp", "trans protocol must be tcp or udp");

    this._netthread = new NetThread(this, this._endpoint, this._protocol);
    this._netthread.initialize();
};

BindAdapter.prototype.stop = function () {
    this._netthread.stop();
};

//当传输端口上获得一个完整的请求包之后
BindAdapter.prototype.dispatch = function ($Transceiver, $protoMessage) {
    var current = new JceCurrent();
    current._origin      = $protoMessage.origin;
    current._endpoint    = new Endpoint($Transceiver.socket.remoteAddress, $Transceiver.socket.remotePort);
    current._adapter     = this;
    current._transceiver = $Transceiver;
    current._timestamp   = TimeProvider.nowTimestamp();

    try {
        if (this.isTafProtocol() === true) {
            //处理TAF协议
            var iReturn = this._handleImp.onDispatch(current, $protoMessage.sFuncName, $protoMessage.origin.sBuffer);
            if (iReturn != TafError.SUCCESS) {
                current.sendErrorResponse(iReturn, "");
            }
        } else {
            //处理非TAF协议
            var Func = ($protoMessage.sFuncName && this.handleImp[$protoMessage.sFuncName]) || this.handleImp["doRequest"];
            if (Func) {
                Func.call(this.handleImp, current, $protoMessage.origin);
            } else {
                current.sendErrorResponse(TafError.SERVER.FUNC_NOT_FOUND, "Not Found Func By Name:" + ($protoMessage.sFuncName?$protoMessage.sFuncName:"doRequest"));
            }
        }
    } catch (e) {
        if ("number" === typeof e.code) {
            current.sendErrorResponse(e.code, "BindAdapter Exception:" + e.message);
        } else {
            current.sendErrorResponse(TafError.SERVER.DECODE_ERROR, "BindAdapter Exception:" + e.message);
            console.error("[TAF][BindAdapter::dispatch, BindAdapter Exception:", e.message);
        }
    }
};
