// @tencent/taf-rpc/dev/benchmark/test/server/NodeJsCommImp.js
// **********************************************************************
// This file was generated by a TAF parser!
// TAF version 3.1.1.2 by WSRD Tencent.
// Generated from `publish.jce'
// **********************************************************************

var TRom = require('./NodeJsComm.js').TRom;
module.exports.TRom = TRom;

TRom.NodeJsCommImp.prototype.initialize = function ( ) {
    //TODO::

}

TRom.NodeJsCommImp.prototype.getPlayer = function (current, sUserName, p) {
    //TODO::

}

TRom.NodeJsCommImp.prototype.getUsrName = function (current, sUsrName, sValue1, sValue2) {
    //TODO::
    sValue1 = "OUT_" + sUsrName + "_1";
    sValue2 = "OUT_" + sUsrName + "_2";
    current.sendResponse(100, sValue1, sValue2);
}

TRom.NodeJsCommImp.prototype.testInt = function (current, inInt) {
    console.log("IN:", inInt);
    current.sendResponse(0, inInt);
}

TRom.NodeJsCommImp.prototype.getall = function (current, stUser, stResult) {
    //TODO::

}

TRom.NodeJsCommImp.prototype.test = function (current) {
    //TODO::

}
