// @tencent/taf-rpc/dev/taf-rpc/demo.1/server.node.1/NodeJsCommImp.js
// **********************************************************************
// This file was generated by a TAF parser!
// TAF version 3.1.1.2 by WSRD Tencent.
// Generated from `publish.jce'
// **********************************************************************

var TRom = require('./NodeJsComm.js').TRom;
module.exports.TRom = TRom;

TRom.NodeJsCommImp.prototype.initialize = function ( ) {
    //TODO::
    //this.aPrx = this.application.getCommunicator().stringToProxy(TRom, "TRom.APproxy.TestObj@127.0.0.1");
    console.log("NodeJsCommImp::initialize");
};

TRom.NodeJsCommImp.prototype.test = function (current) {
    //TODO::

};

TRom.NodeJsCommImp.prototype.getall = function (current, stUser, stResult) {
    //TODO::
    console.log("stUser.name::", stUser.name);

    stResult.iLevel = 11000;
};

var index = 0;

TRom.NodeJsCommImp.prototype.getUsrName = function (current, sUsrName, sValue1, sValue2) {
    //TODO::
    console.log("sUsrName::", sUsrName);
    sValue1 = "OUT_" + sUsrName + "_1";
    sValue2 = "OUT_" + sUsrName + "_2";

    current.sendResponse(100002, sValue1, sValue2);

    if (++index == 5) {
        console.log("to stop");
        current.application.stop();
    }
};
