// @tencent/taf-rpc/examples/taf-rpc/demo.1/client.node.wup.socket/NodeTafJce.js
// **********************************************************************
// This file was generated by a TAF parser!
// TAF version 3.1.1.2 by WSRD Tencent.
// Generated from `./NodeTaf.jce'
// **********************************************************************

var TafStream = require('../../../../../taf-stream');
var TafError  = require("../../../../protal.js").error;

var TRom = TRom || {};
module.exports.TRom = TRom;

TRom.NodeTafProxy = function () {
    this._name   = undefined;
    this._worker = undefined;
}

TRom.NodeTafProxy.prototype.setTimeout = function (iTimeout) {
    this._worker.timeout = iTimeout;
}

TRom.NodeTafProxy.prototype.getTimeout = function ( ) {
    return this._worker.timeout;
}


TRom.User_t = function() {
    this.id = 0;
    this.score = 0;
    this.name = "";
    this._classname = "TRom.User_t";
};
TRom.User_t._classname = "TRom.User_t";

TRom.User_t._write = function (os, tag, value) { os.writeStruct(tag, value); }
TRom.User_t._read  = function (is, tag, def) { return is.readStruct(tag, true, def); }
TRom.User_t._readFrom = function (is) {
    var tmp = new TRom.User_t();
    tmp.id = is.readInt32(0, false, 0);
    tmp.score = is.readInt32(1, false, 0);
    tmp.name = is.readString(2, false, "");
    return tmp;
};
TRom.User_t.prototype._writeTo = function (os) {
    os.writeInt32(0, this.id);
    os.writeInt32(1, this.score);
    os.writeString(2, this.name);
};
TRom.User_t.prototype._equal = function (anItem) {
    assert(false, 'this structure not define key operation');
}
TRom.User_t.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = 'STRUCT' + Math.random();
    }
    return this._proto_struct_name_;
}
TRom.User_t.prototype.toObject = function() { 
    var tmp = {};

    tmp.id = this.id;
    tmp.score = this.score;
    tmp.name = this.name;
    
    return tmp;
}
TRom.User_t.prototype.readFromObject = function(json) { 
    !json.hasOwnProperty("id") || (this.id = json.id);
    !json.hasOwnProperty("score") || (this.score = json.score);
    !json.hasOwnProperty("name") || (this.name = json.name);
}
TRom.User_t.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
}
TRom.User_t.new = function () {
    return new TRom.User_t();
}
TRom.User_t.create = function (is) {
    return TRom.User_t._readFrom(is);
}
TRom.Result_t = function() {
    this.id = 0;
    this.iLevel = 0;
    this._classname = "TRom.Result_t";
};
TRom.Result_t._classname = "TRom.Result_t";

TRom.Result_t._write = function (os, tag, value) { os.writeStruct(tag, value); }
TRom.Result_t._read  = function (is, tag, def) { return is.readStruct(tag, true, def); }
TRom.Result_t._readFrom = function (is) {
    var tmp = new TRom.Result_t();
    tmp.id = is.readInt32(0, false, 0);
    tmp.iLevel = is.readInt32(1, false, 0);
    return tmp;
};
TRom.Result_t.prototype._writeTo = function (os) {
    os.writeInt32(0, this.id);
    os.writeInt32(1, this.iLevel);
};
TRom.Result_t.prototype._equal = function (anItem) {
    assert(false, 'this structure not define key operation');
}
TRom.Result_t.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = 'STRUCT' + Math.random();
    }
    return this._proto_struct_name_;
}
TRom.Result_t.prototype.toObject = function() { 
    var tmp = {};

    tmp.id = this.id;
    tmp.iLevel = this.iLevel;
    
    return tmp;
}
TRom.Result_t.prototype.readFromObject = function(json) { 
    !json.hasOwnProperty("id") || (this.id = json.id);
    !json.hasOwnProperty("iLevel") || (this.iLevel = json.iLevel);
}
TRom.Result_t.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
}
TRom.Result_t.new = function () {
    return new TRom.Result_t();
}
TRom.Result_t.create = function (is) {
    return TRom.Result_t._readFrom(is);
}
