// @tencent/taf-config/ConfigProxy.js
// **********************************************************************
// Parsed By TAFParser(3.2.2.11), Generated By Jce2Node(20180425)
// TAFParser Maintained By <WSRD> and Jce2Node Maintained By <superzheng>
// Generated from "Config.jce" by Client Mode
// **********************************************************************

"use strict";

var assert    = require("assert");
var TafStream = require("@tencent/taf-stream");
var TafError  = require("@tencent/taf-rpc").error;

var taf = taf || {};
module.exports.taf = taf;

taf.ConfigProxy = function () {
    this._name   = undefined;
    this._worker = undefined;
};

taf.ConfigProxy.prototype.setTimeout = function (iTimeout) {
    this._worker.timeout = iTimeout;
};

taf.ConfigProxy.prototype.getTimeout = function () {
    return this._worker.timeout;
};


taf.ConfigInfo = function() {
    this.appname = "";
    this.servername = "";
    this.filename = "";
    this.bAppOnly = false;
    this.host = "";
    this.setdivision = "";
    this.containername = "";
    this._classname = "taf.ConfigInfo";
};
taf.ConfigInfo._classname = "taf.ConfigInfo";
taf.ConfigInfo._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.ConfigInfo._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.ConfigInfo._readFrom = function (is) {
    var tmp = new taf.ConfigInfo();
    tmp.appname = is.readString(0, true, "");
    tmp.servername = is.readString(1, true, "");
    tmp.filename = is.readString(2, true, "");
    tmp.bAppOnly = is.readBoolean(3, true, false);
    tmp.host = is.readString(4, false, "");
    tmp.setdivision = is.readString(5, false, "");
    tmp.containername = is.readString(6, false, "");
    return tmp;
};
taf.ConfigInfo.prototype._writeTo = function (os) {
    os.writeString(0, this.appname);
    os.writeString(1, this.servername);
    os.writeString(2, this.filename);
    os.writeBoolean(3, this.bAppOnly);
    os.writeString(4, this.host);
    os.writeString(5, this.setdivision);
    os.writeString(6, this.containername);
};
taf.ConfigInfo.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.ConfigInfo.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.ConfigInfo.prototype.toObject = function() { 
    return {
        "appname" : this.appname,
        "servername" : this.servername,
        "filename" : this.filename,
        "bAppOnly" : this.bAppOnly,
        "host" : this.host,
        "setdivision" : this.setdivision,
        "containername" : this.containername
    };
};
taf.ConfigInfo.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("appname") && (this.appname = json.appname);
    json.hasOwnProperty("servername") && (this.servername = json.servername);
    json.hasOwnProperty("filename") && (this.filename = json.filename);
    json.hasOwnProperty("bAppOnly") && (this.bAppOnly = json.bAppOnly);
    json.hasOwnProperty("host") && (this.host = json.host);
    json.hasOwnProperty("setdivision") && (this.setdivision = json.setdivision);
    json.hasOwnProperty("containername") && (this.containername = json.containername);
};
taf.ConfigInfo.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.ConfigInfo.new = function () {
    return new taf.ConfigInfo();
};
taf.ConfigInfo.create = function (is) {
    return taf.ConfigInfo._readFrom(is);
};

taf.GetConfigListInfo = function() {
    this.appname = "";
    this.servername = "";
    this.bAppOnly = false;
    this.host = "";
    this.setdivision = "";
    this.containername = "";
    this._classname = "taf.GetConfigListInfo";
};
taf.GetConfigListInfo._classname = "taf.GetConfigListInfo";
taf.GetConfigListInfo._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.GetConfigListInfo._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.GetConfigListInfo._readFrom = function (is) {
    var tmp = new taf.GetConfigListInfo();
    tmp.appname = is.readString(0, true, "");
    tmp.servername = is.readString(1, false, "");
    tmp.bAppOnly = is.readBoolean(2, false, false);
    tmp.host = is.readString(3, false, "");
    tmp.setdivision = is.readString(4, false, "");
    tmp.containername = is.readString(5, false, "");
    return tmp;
};
taf.GetConfigListInfo.prototype._writeTo = function (os) {
    os.writeString(0, this.appname);
    os.writeString(1, this.servername);
    os.writeBoolean(2, this.bAppOnly);
    os.writeString(3, this.host);
    os.writeString(4, this.setdivision);
    os.writeString(5, this.containername);
};
taf.GetConfigListInfo.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.GetConfigListInfo.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.GetConfigListInfo.prototype.toObject = function() { 
    return {
        "appname" : this.appname,
        "servername" : this.servername,
        "bAppOnly" : this.bAppOnly,
        "host" : this.host,
        "setdivision" : this.setdivision,
        "containername" : this.containername
    };
};
taf.GetConfigListInfo.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("appname") && (this.appname = json.appname);
    json.hasOwnProperty("servername") && (this.servername = json.servername);
    json.hasOwnProperty("bAppOnly") && (this.bAppOnly = json.bAppOnly);
    json.hasOwnProperty("host") && (this.host = json.host);
    json.hasOwnProperty("setdivision") && (this.setdivision = json.setdivision);
    json.hasOwnProperty("containername") && (this.containername = json.containername);
};
taf.GetConfigListInfo.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.GetConfigListInfo.new = function () {
    return new taf.GetConfigListInfo();
};
taf.GetConfigListInfo.create = function (is) {
    return taf.GetConfigListInfo._readFrom(is);
};


var __taf_Config$ListAllConfigByInfo$EN = function (configInfo) {
    var os = new TafStream.JceOutputStream();
    os.writeStruct(1, configInfo);
    return os.getBinBuffer();
};

var __taf_Config$ListAllConfigByInfo$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0),
                "arguments" : {
                    "vf" : is.readList(2, true, TafStream.List(TafStream.String))
                }
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_Config$ListAllConfigByInfo$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.ConfigProxy.prototype.ListAllConfigByInfo = function (configInfo) {
    return this._worker.taf_invoke("ListAllConfigByInfo", __taf_Config$ListAllConfigByInfo$EN(configInfo), arguments[arguments.length - 1]).then(__taf_Config$ListAllConfigByInfo$DE, __taf_Config$ListAllConfigByInfo$ER);
};

var __taf_Config$ListConfig$EN = function (app, server) {
    var os = new TafStream.JceOutputStream();
    os.writeString(1, app);
    os.writeString(2, server);
    return os.getBinBuffer();
};

var __taf_Config$ListConfig$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0),
                "arguments" : {
                    "vf" : is.readList(3, true, TafStream.List(TafStream.String))
                }
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_Config$ListConfig$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.ConfigProxy.prototype.ListConfig = function (app, server) {
    return this._worker.taf_invoke("ListConfig", __taf_Config$ListConfig$EN(app, server), arguments[arguments.length - 1]).then(__taf_Config$ListConfig$DE, __taf_Config$ListConfig$ER);
};

var __taf_Config$ListConfigByInfo$EN = function (configInfo) {
    var os = new TafStream.JceOutputStream();
    os.writeStruct(1, configInfo);
    return os.getBinBuffer();
};

var __taf_Config$ListConfigByInfo$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0),
                "arguments" : {
                    "vf" : is.readList(2, true, TafStream.List(TafStream.String))
                }
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_Config$ListConfigByInfo$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.ConfigProxy.prototype.ListConfigByInfo = function (configInfo) {
    return this._worker.taf_invoke("ListConfigByInfo", __taf_Config$ListConfigByInfo$EN(configInfo), arguments[arguments.length - 1]).then(__taf_Config$ListConfigByInfo$DE, __taf_Config$ListConfigByInfo$ER);
};

var __taf_Config$checkConfig$EN = function (appServerName, filename, host) {
    var os = new TafStream.JceOutputStream();
    os.writeString(1, appServerName);
    os.writeString(2, filename);
    os.writeString(3, host);
    return os.getBinBuffer();
};

var __taf_Config$checkConfig$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0),
                "arguments" : {
                    "result" : is.readString(4, true, "")
                }
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_Config$checkConfig$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.ConfigProxy.prototype.checkConfig = function (appServerName, filename, host) {
    return this._worker.taf_invoke("checkConfig", __taf_Config$checkConfig$EN(appServerName, filename, host), arguments[arguments.length - 1]).then(__taf_Config$checkConfig$DE, __taf_Config$checkConfig$ER);
};

var __taf_Config$checkConfigByInfo$EN = function (configInfo) {
    var os = new TafStream.JceOutputStream();
    os.writeStruct(1, configInfo);
    return os.getBinBuffer();
};

var __taf_Config$checkConfigByInfo$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0),
                "arguments" : {
                    "result" : is.readString(2, true, "")
                }
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_Config$checkConfigByInfo$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.ConfigProxy.prototype.checkConfigByInfo = function (configInfo) {
    return this._worker.taf_invoke("checkConfigByInfo", __taf_Config$checkConfigByInfo$EN(configInfo), arguments[arguments.length - 1]).then(__taf_Config$checkConfigByInfo$DE, __taf_Config$checkConfigByInfo$ER);
};

var __taf_Config$loadConfig$EN = function (app, server, filename) {
    var os = new TafStream.JceOutputStream();
    os.writeString(1, app);
    os.writeString(2, server);
    os.writeString(3, filename);
    return os.getBinBuffer();
};

var __taf_Config$loadConfig$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0),
                "arguments" : {
                    "config" : is.readString(4, true, "")
                }
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_Config$loadConfig$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.ConfigProxy.prototype.loadConfig = function (app, server, filename) {
    return this._worker.taf_invoke("loadConfig", __taf_Config$loadConfig$EN(app, server, filename), arguments[arguments.length - 1]).then(__taf_Config$loadConfig$DE, __taf_Config$loadConfig$ER);
};

var __taf_Config$loadConfigByHost$EN = function (appServerName, filename, host) {
    var os = new TafStream.JceOutputStream();
    os.writeString(1, appServerName);
    os.writeString(2, filename);
    os.writeString(3, host);
    return os.getBinBuffer();
};

var __taf_Config$loadConfigByHost$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0),
                "arguments" : {
                    "config" : is.readString(4, true, "")
                }
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_Config$loadConfigByHost$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.ConfigProxy.prototype.loadConfigByHost = function (appServerName, filename, host) {
    return this._worker.taf_invoke("loadConfigByHost", __taf_Config$loadConfigByHost$EN(appServerName, filename, host), arguments[arguments.length - 1]).then(__taf_Config$loadConfigByHost$DE, __taf_Config$loadConfigByHost$ER);
};

var __taf_Config$loadConfigByInfo$EN = function (configInfo) {
    var os = new TafStream.JceOutputStream();
    os.writeStruct(1, configInfo);
    return os.getBinBuffer();
};

var __taf_Config$loadConfigByInfo$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0),
                "arguments" : {
                    "config" : is.readString(2, true, TafStream.BinBuffer.from("").toNodeBufferUnSafe(), 1)
                }
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_Config$loadConfigByInfo$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.ConfigProxy.prototype.loadConfigByInfo = function (configInfo) {
    return this._worker.taf_invoke("loadConfigByInfo", __taf_Config$loadConfigByInfo$EN(configInfo), arguments[arguments.length - 1]).then(__taf_Config$loadConfigByInfo$DE, __taf_Config$loadConfigByInfo$ER);
};
