// @tencent/taf-config/index.js
"use strict";

var util = require("util");
var assert = require("assert");
var iconv = require("iconv-lite");
var EventEmitter = require("events").EventEmitter;

var Promise = require("@tencent/taf-utils").Promise;
var ConfigParser = require("@tencent/taf-utils").Config;
var TafClient = require("@tencent/taf-rpc").Communicator.New();

var taf = require("./ConfigProxy").taf;

var FORMAT = {
	'C' : 'C',
	'JSON' : 'JSON',
	'TEXT' : 'TEXT'
};
var LOCATION = {
    'SERVER' : 'SERVER',
    'APP' : 'APP'
};
var CONFIG_OBJ = "taf.tafconfig.ConfigObj";
var TIMEOUT = 30 * 1000;

function TafConfig(obj) {
	var parser;

	obj = obj || process.env.TAF_CONFIG;

	assert(obj, 'TAF_CONFIG is not in env and init argument is neither an Object nor a String.');

	if (typeof obj === 'string' && obj !== '') {
		TafClient.initialize(obj);
		parser = new ConfigParser();
		parser.parseFile(obj);
	} else {
		parser = obj;
	}

	assert(typeof parser === 'object', 'init argument not an TAFCnfigureObj');

	this.ConfigObj = parser.get('taf.application.server.config', CONFIG_OBJ);

	this.app = parser.get('taf.application.server.app');
	assert(this.app, 'app not found in TAF_CONFIG(taf.application.server.app)');

	this.server = parser.get('taf.application.server.server');
	assert(this.server, 'server not found in TAF_CONFIG(taf.application.server.server)');

	this.localip = parser.get('taf.application.server.localip');
	assert(this.localip, 'localip not found in TAF_CONFIG(taf.application.server.localip)');

	this.containername = parser.get('taf.application.server.container');

	if (parser.get('taf.application.enableset', '').toLowerCase() === 'y') {
		this.setdivision = parser.get('taf.application.setdivision');
		assert(this.setdivision, 'setdivision is empty in TAF_CONFIG(taf.application.setdivision)');
	}

	this.timeout = TIMEOUT;

	this.setMaxListeners(0);
}

util.inherits(TafConfig, EventEmitter);

TafConfig.prototype.FORMAT = FORMAT;
TafConfig.prototype.LOCATION = LOCATION;

TafConfig.prototype._init = function() {
	var ths = this;

	this._client = TafClient.stringToProxy(taf.ConfigProxy, this.ConfigObj);
	this._client.setTimeout(this.timeout);

	process.on('message', function(obj) {
		if (obj.cmd === "taf.loadconfig") {
			ths.emit("configPushed", obj.data);
		}
	});
};

TafConfig.prototype.loadServerConfig = function(options) {
	assert(!options || !options.isAppConfig, 'Can\'t load server config in app');

	return this._loadSingleConfig(this.server + ".conf", options);
};

TafConfig.prototype._loadSingleConfig = function(filename, options) {
	var info = new taf.ConfigInfo(), opts = options || {};
	info.appname = this.app;
	info.servername = this.server;
	info.filename = filename;
	info.host = this.localip;
	if (this.setdivision) {
		info.setdivision = this.setdivision;
	}
	if (this.containername) {
		info.containername = this.containername;
	}
	if (LOCATION.hasOwnProperty(opts.location)) {
		info.bAppOnly = (opts.location === LOCATION.APP);
		if (info.bAppOnly) {
			info.servername = '';
		}
	} else {
		info.bAppOnly = false;
	}

	if (!this._client) {
		this._init();
	}

	return this._client.loadConfigByInfo(info).then(function(resp) {
		var content = iconv.decode(resp.response.arguments.config, "GBK"), format = FORMAT.C;

		if (FORMAT.hasOwnProperty(opts.format)) {
			format = opts.format;
		}
		switch (format) {
			case FORMAT.C : {
				var parser = new ConfigParser();
				parser.parseText(content, "utf8");
				return parser.data;
			}
			case FORMAT.JSON : {
				return JSON.parse(content);
			}
			case FORMAT.TEXT : {
				return content;
			}
		}
	});
};

TafConfig.prototype._loadMultiConfig = function(fileList, options) {
	var ths = this;

	return Promise.all(fileList.map(function(filename) {
		return ths.loadConfig(filename, options).then(function(content) {
			return {
				"filename" : filename, 
				"content" : content
			};
		});
	}));
};

TafConfig.prototype.loadList = function(options) {
	var info = new taf.GetConfigListInfo(), opts = options || {};
	info.appname = this.app;
	info.servername = this.server;
	if (this.setdivision) {
		info.setdivision = this.setdivision;
	}
	if (this.containername) {
		info.containername = this.containername;
	}
	if (LOCATION.hasOwnProperty(opts.location)) {
		info.bAppOnly = (opts.location === LOCATION.APP);
		if (info.bAppOnly) {
			info.servername = '';
		}
	} else {
		info.bAppOnly = false;
	}

	if (!this._client) {
		this._init();
	}

	return this._client.ListAllConfigByInfo(info).then(function(resp) {
		return resp.response.arguments.vf.value;
	});
};

TafConfig.prototype.loadConfig = function(fileList, options) {
	var ths = this, _options = null;

	if (Array.isArray(fileList)) {
		return this._loadMultiConfig(fileList, options);
	} else if (typeof fileList === 'string' && fileList !== '') {
		return this._loadSingleConfig(fileList, options);
	} else {
		_options = fileList;
		assert(!_options || !_options.isAppConfig, 'Can\'t load all app config');
		return this.loadList(_options).then(function(_fileList) {
			return ths._loadMultiConfig(_fileList, _options);
		});
	}
};

module.exports = exports = TafConfig;
