// @tencent/taf-rpc/dev/taf-rpc/demo.1/server.node.2/NodeJsCommImp.js
// **********************************************************************
// This file was generated by a TAF parser!
// TAF version 3.1.1.2 by WSRD Tencent.
// Generated from `publish.jce'
// **********************************************************************

var TRom = require('./NodeJsComm.js').TRom;
module.exports.TRom = TRom;

TRom.NodeJsCommImp.prototype.initialize = function ( ) {
    //TODO::
    //this.aPrx = this.application.getCommunicator().stringToProxy(TRom, "TRom.APproxy.TestObj@127.0.0.1");
    console.log("NodeJsCommImp::initialize");
}

TRom.NodeJsCommImp.prototype.test = function (current) {
    //TODO::

}

TRom.NodeJsCommImp.prototype.getall = function (current, stUser, stResult) {
    //TODO::
    console.log("stUser.name::", stUser.name);

    stResult.iLevel = 11000;
}

TRom.NodeJsCommImp.prototype.getUsrName = function (current, sUsrName, sValue1, sValue2) {
    //可以获取的请求数据
    console.log("\nTRom.NodeJsCommImp.prototype.getUsrName.......");
    console.log("client endpoint:", current.getEndpoint().toString());

    var context = current.getContext();
    var status  = current.getRequestStatus();
    context.forEach(function (key, value){
        console.log("context:", key, value);
    });

    console.log("RequestId:", current.getRequestId());

    //构造返回的数据
    sValue1 = "svr_" + sUsrName + "_1";
    sValue2 = "svr_" + sUsrName + "_2";

    current.setResponseContext({a:1, b:2});
    current.sendResponse(100, sValue1, sValue2);
}
