// @tencent/taf-rpc/core/server/JceCurrent.js
var TafDyeing    = require("@tencent/taf-dyeing");
var TimeProvider = require("@tencent/taf-utils").timeProvider;
var TafMonitor   = require("@tencent/taf-monitor");
var ProtoMessage = require("./ProtoMessage.js");
var Configure    = require("@tencent/taf-utils").Config;


///////////////////////////////////////////////程序运行的上下文信息/////////////////////////////////////////////////////
var JceCurrent = function () {
    this._origin         = undefined;    //原始的请求结构体
    this._endpoint       = undefined;    //客户端的IP地址
    this._adapter        = undefined;    //BindAdapter的指向
    this._servant        = undefined;    //处理类
    this._transceiver    = undefined;    //连接
    this._timestamp      = 0;            //接受到完整请求包的时间
    this._rspcontext     = {};
    this._configure      = undefined;    //配置
};
module.exports.JceCurrent = JceCurrent;

/***
 * 对外提供的获取接口
 ***/
JceCurrent.prototype.__defineGetter__("application",  function () { return this._adapter.handleImp.application; });

JceCurrent.prototype.__defineGetter__("communicator", function () { return this._adapter.handleImp.application.communicator; });

//获取客户端的IP和端口
JceCurrent.prototype.getEndpoint = function () {
    return this._endpoint;
};

//仅TAF协议有效.请求的协议的版本号
JceCurrent.prototype.getRequestVersion = function () {
    return this._origin?this._origin.iVersion:undefined;
};

//请求ID
JceCurrent.prototype.getRequestId = function () {
    return this._origin?this._origin.iRequestId:undefined;
};

//仅TAF协议有效.获取扩展map
JceCurrent.prototype.getContext = function () {
    return this._origin?this._origin.context:undefined;
};

//仅TAF协议有效.保存状态信息，比如灰度、染色等
JceCurrent.prototype.getRequestStatus = function () {
    return this._origin?this._origin.status:undefined;
};

//设置返回的context
JceCurrent.prototype.setResponseContext = function ($context) {
    this._rspcontext = $context;
};

//获取原始的请求数据
JceCurrent.prototype.getRequest = function () {
    return this._origin;
};

//适用于第三份协议时的回包函数
JceCurrent.prototype.sendResponse = function () {
    var response = new ProtoMessage.Response();
    response.origin       = this._origin;
    response.appBuffer    = Array.prototype.slice.call(arguments, 0);
    response.iResultCode  = 0;
    response.sResultDesc  = "";

    this._transceiver.send(response);
};

// 仅TAF协议有效.开发者调用sendResponse，编码之后，由框架调用该函数返回数据
JceCurrent.prototype.doResponse = function ($BinBuffer) {
    // 处理三种需要主动上报的数据[wup_client,one_way_client,not_taf_client]
    this._report();

    // 如果是TAF协议并且是单向调用则直接返回
    if (this._adapter.isTafProtocol() && this._origin.cPacketType == 1) { //1表示JCEONEWAY，单向调用
        return ;
    }

    // 生成一个向协议解析器传递的结构体
    var response = new ProtoMessage.Response();
    response.origin       = this._origin;
    response.appBuffer    = $BinBuffer;
    response.iResultCode  = 0;
    response.sResultDesc  = "";
    response.property     = {context:this._rspcontext};

    this._transceiver.send(response);
};

// 当出现错误时，需要给个客户端返回一个错误消息是调用
JceCurrent.prototype.sendErrorResponse = function ($iRet, $sMessage) {
    // 如果是TAF协议并且是单向调用则直接返回
    if (this._adapter.isTafProtocol() && this._origin.cPacketType == 1) { //1表示JCEONEWAY，单向调用
        return ;
    }

    // 生成一个向协议解析器传递的结构体
    var response = new ProtoMessage.Response();
    response.origin       = this._origin;
    response.appBuffer    = undefined;
    response.iResultCode  = $iRet;
    response.sResultDesc  = $sMessage;

    this._transceiver.send(response);
};

//获取原始请求的染色情况
JceCurrent.prototype.getDyeingObj = function () {
    var bDyeing    = this._origin.iMessageType & 0x04;
    var sDyeingKey = this._origin.status.get("STATUS_DYED_KEY") || '';

    return TafDyeing.gen(bDyeing, sDyeingKey);
};

//获取taf鉴权后设置的userid
JceCurrent.prototype.getUserId = function(){
    return this._transceiver.getUserId();
}

// 处理三种需要主动上报的数据[wup_client,one_way_client,not_taf_client]
JceCurrent.prototype._report = function () {
    // 判断是否在测试环境中，非正式环境不上报
    if (process.env.TAF_MONITOR === undefined) {
        return ;
    }
    if (this.communicator === undefined) { //如果服务端没有初始化通信器的情况下，则不上报
        return ;
    }

    var sReportType = "";
    if (this._origin.iVersion === 2 || this._origin.iVersion === 3) { //如果是wup协议调用
        sReportType = "wup_client";
    } else if (this._origin.cPacketType === 1) { //如果是单向调用
        sReportType = "one_way_client";
    } else if (this._adapter.isTafProtocol() === false) { //如果是第三方调用方式
        sReportType = "not_taf_client";
    } else { //除上述三种情况之外，直接返回
        return ;
    }

    var slaveContainer = '', slaveSetName = '', slaveSetArea = '', slaveSetID = '', slaveSetString = '', slaveName = '';

    if(process.env.TAF_CONFIG)
    {
        // 读取配置文件
        if(!this._configure)
        {
            this._configure = new Configure();
            this._configure.parseFile(process.env.TAF_CONFIG);
        }
    }

    if(this._configure)
    {
        var enableSet = this._configure.get("taf.application.enableset", 'n');
        var isDocker = this._configure.get("taf.application.isdocker", 'n');
        slaveName = this._configure.get("taf.application.client.modulename", '');

        if(enableSet == 'y' || enableSet == 'Y')
        {
            var setDivision = this._configure.get("taf.application.setdivision", undefined);
            if(setDivision)
            {
                slaveSetString = setDivision;
                var arr = slaveSetString.split('.');
                slaveSetName = arr[0] || '';
                slaveSetArea = arr[1] || '';
                slaveSetID = arr[2] || '';

                slaveName += '.' + arr.join('');
            }
        }

        if(isDocker == 'y' || isDocker == 'Y')
        {
            var container = this._configure.get("taf.application.container", undefined);
            if(container)
                slaveContainer = container;
        }
    }

    var headers = {
        masterName    : sReportType,
        slaveName     : slaveName,
        interfaceName : this._origin.sFuncName,
        masterIp      : this._endpoint.sHost,
        slaveIp       : "",
        slavePort     : 0,
        bFromClient   : false
    };

    if(slaveSetName)
        headers.slaveSetName = slaveSetName;

    if(slaveSetArea)
        headers.slaveSetArea = slaveSetArea;

    if(slaveSetID)
        headers.slaveSetID = slaveSetID;

    if(slaveContainer)
        headers.slaveContainer = slaveContainer;

    TafMonitor.stat.report(headers, TafMonitor.stat.TYPE.SUCCESS, TimeProvider.nowTimestamp() - this._timestamp);
};
