// @tencent/taf-registry/EndpointFJce.js
// **********************************************************************
// This file was generated by a TAF parser!
// TAF version 3.2.1.6 by WSRD Tencent.
// Generated from `./EndpointF.jce'
// **********************************************************************

var assert    = require('assert');
var TafStream = require('@tencent/taf-stream');


var taf = taf || {};
module.exports.taf = taf;

taf.EndpointF = function() {
    this.host = "";
    this.port = 0;
    this.timeout = 0;
    this.istcp = 0;
    this.grid = 0;
    this.groupworkid = 0;
    this.grouprealid = 0;
    this.setId = "";
    this.qos = 0;
    this.bakFlag = 0;
    this.gridFlag = 0;
    this.weight = 0;
    this.weightType = 0;
    this.cpuload = 0;
    this.sampletime = 0;
    this.containerName = "";
    this._classname = "taf.EndpointF";
};
taf.EndpointF._classname = "taf.EndpointF";

taf.EndpointF._write = function (os, tag, value) { os.writeStruct(tag, value); }
taf.EndpointF._read  = function (is, tag, def) { return is.readStruct(tag, true, def); }
taf.EndpointF._readFrom = function (is) {
    var tmp = new taf.EndpointF();
    tmp.host = is.readString(0, true, "");
    tmp.port = is.readInt32(1, true, 0);
    tmp.timeout = is.readInt32(2, true, 0);
    tmp.istcp = is.readInt32(3, true, 0);
    tmp.grid = is.readInt32(4, true, 0);
    tmp.groupworkid = is.readInt32(5, false, 0);
    tmp.grouprealid = is.readInt32(6, false, 0);
    tmp.setId = is.readString(7, false, "");
    tmp.qos = is.readInt32(8, false, 0);
    tmp.bakFlag = is.readInt32(9, false, 0);
    tmp.gridFlag = is.readInt32(10, false, 0);
    tmp.weight = is.readInt32(11, false, 0);
    tmp.weightType = is.readInt32(12, false, 0);
    tmp.cpuload = is.readInt32(13, false, 0);
    tmp.sampletime = is.readInt64(14, false, 0);
    tmp.containerName = is.readString(15, false, "");
    return tmp;
};
taf.EndpointF.prototype._writeTo = function (os) {
    os.writeString(0, this.host);
    os.writeInt32(1, this.port);
    os.writeInt32(2, this.timeout);
    os.writeInt32(3, this.istcp);
    os.writeInt32(4, this.grid);
    os.writeInt32(5, this.groupworkid);
    os.writeInt32(6, this.grouprealid);
    os.writeString(7, this.setId);
    os.writeInt32(8, this.qos);
    os.writeInt32(9, this.bakFlag);
    os.writeInt32(10, this.gridFlag);
    os.writeInt32(11, this.weight);
    os.writeInt32(12, this.weightType);
    os.writeInt32(13, this.cpuload);
    os.writeInt64(14, this.sampletime);
    os.writeString(15, this.containerName);
};
taf.EndpointF.prototype._equal = function (anItem) {
    return this.host === anItem.host && 
        this.port === anItem.port && 
        this.timeout === anItem.timeout && 
        this.istcp === anItem.istcp && 
        this.grid === anItem.grid && 
        this.qos === anItem.qos && 
        this.gridFlag === anItem.gridFlag && 
        this.weight === anItem.weight && 
        this.weightType === anItem.weightType;
}
taf.EndpointF.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = 'STRUCT' + Math.random();
    }
    return this._proto_struct_name_;
}
taf.EndpointF.prototype.toObject = function() { 
    var tmp = {};

    tmp.host = this.host;
    tmp.port = this.port;
    tmp.timeout = this.timeout;
    tmp.istcp = this.istcp;
    tmp.grid = this.grid;
    tmp.groupworkid = this.groupworkid;
    tmp.grouprealid = this.grouprealid;
    tmp.setId = this.setId;
    tmp.qos = this.qos;
    tmp.bakFlag = this.bakFlag;
    tmp.gridFlag = this.gridFlag;
    tmp.weight = this.weight;
    tmp.weightType = this.weightType;
    tmp.cpuload = this.cpuload;
    tmp.sampletime = this.sampletime;
    tmp.containerName = this.containerName;
    
    return tmp;
}
taf.EndpointF.prototype.readFromObject = function(json) { 
    !json.hasOwnProperty("host") || (this.host = json.host);
    !json.hasOwnProperty("port") || (this.port = json.port);
    !json.hasOwnProperty("timeout") || (this.timeout = json.timeout);
    !json.hasOwnProperty("istcp") || (this.istcp = json.istcp);
    !json.hasOwnProperty("grid") || (this.grid = json.grid);
    !json.hasOwnProperty("groupworkid") || (this.groupworkid = json.groupworkid);
    !json.hasOwnProperty("grouprealid") || (this.grouprealid = json.grouprealid);
    !json.hasOwnProperty("setId") || (this.setId = json.setId);
    !json.hasOwnProperty("qos") || (this.qos = json.qos);
    !json.hasOwnProperty("bakFlag") || (this.bakFlag = json.bakFlag);
    !json.hasOwnProperty("gridFlag") || (this.gridFlag = json.gridFlag);
    !json.hasOwnProperty("weight") || (this.weight = json.weight);
    !json.hasOwnProperty("weightType") || (this.weightType = json.weightType);
    !json.hasOwnProperty("cpuload") || (this.cpuload = json.cpuload);
    !json.hasOwnProperty("sampletime") || (this.sampletime = json.sampletime);
    !json.hasOwnProperty("containerName") || (this.containerName = json.containerName);
}
taf.EndpointF.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
}
taf.EndpointF.new = function () {
    return new taf.EndpointF();
}
taf.EndpointF.create = function (is) {
    return taf.EndpointF._readFrom(is);
}
