// @tencent/taf-monitor/property/index.js
'use strict';

var assert = require('assert');

var Taf = require('@tencent/taf-rpc');
var TafConfigure = require('@tencent/taf-config-parser');
var TafStream = require('@tencent/taf-stream');

var PropertyF = require('./PropertyFProxy');

exports.PropertyObj = 'taf.tafproperty.PropertyObj';
exports.reportInterval = 10 * 1000;

var emptyfn = function() {};

var data = {}, timer_id, client, msgHeaders;

var ReportObj = function(headers, policies) {
	var ths = this;

	this.policies = Array.isArray(policies) ? policies : [policies];
	this.StatPropMsgHead = new PropertyF.taf.StatPropMsgHead();
	Object.getOwnPropertyNames(headers).forEach(function(name) {
		ths.StatPropMsgHead[name] = headers[name];
	});
};
ReportObj.prototype.report = function(value) {
	this.policies.forEach(function(policy) {
		policy.add(value);
	});

	if (!timer_id) {
		timer_id = setTimeout(task, exports.reportInterval);
	}
};
ReportObj.prototype.get = function() {
	var StatPropMsgBody = new PropertyF.taf.StatPropMsgBody();

	this.policies.forEach(function(policy) {
		var value = policy.get(), StatPropInfo;
		if (value !== undefined) {
			StatPropInfo = new PropertyF.taf.StatPropInfo();
			StatPropInfo.policy = policy.name;
			StatPropInfo.value = value.toString();
			StatPropMsgBody.vInfo.push(StatPropInfo);
		}
	});

	if (StatPropMsgBody.vInfo.length > 0) {
		return {
			StatPropMsgHead : this.StatPropMsgHead,
			StatPropMsgBody : StatPropMsgBody
		};
	} else {
		return null;
	}
};

var task = function() {
	var statmsg = new TafStream.Map(PropertyF.taf.StatPropMsgHead, PropertyF.taf.StatPropMsgBody);

	if (!client) {
		client = Taf.client.stringToProxy(PropertyF.taf.PropertyFProxy, exports.PropertyObj);
	}

	Object.getOwnPropertyNames(data).map(function(key) {
		var obj = data[key].get();
		if (obj) {
			statmsg.put(obj.StatPropMsgHead, obj.StatPropMsgBody);
		}
	});

	if (statmsg.size() > 0) {
		client.reportPropMsg(statmsg, {packetType : 1}).catch(emptyfn);
	}

	timer_id = undefined;
};

var init = function(obj) {
	var setdivision, container, tafConfig;

	obj = obj || process.env.TAF_CONFIG;

	assert(obj, 'TAF_CONFIG is not in env and init argument is neither an Object nor a String.');
	
	if (typeof obj === 'string' && obj !== '') {
		Taf.client.initialize(obj);
		tafConfig = new TafConfigure();
		tafConfig.parseFile(obj);
	} else {
		tafConfig = obj;
	}

	if (typeof tafConfig === 'object') {
		exports.PropertyObj = tafConfig.get('taf.application.client.property', exports.PropertyObj);

		if (!isNaN(parseInt(tafConfig.get('taf.application.client.report-interval')))) {
			exports.reportInterval = parseInt(tafConfig.get('taf.application.client.report-interval'));
		}

		msgHeaders = {
			moduleName : tafConfig.get('taf.application.client.modulename', 'NO_MODULE_NAME'),
			ip : tafConfig.get('taf.application.server.localip', '127.0.0.1'),
			iPropertyVer : 2
		};

		setdivision = tafConfig.get('taf.application.setdivision');

		if (tafConfig.get('taf.application.enableset', '').toLowerCase() === 'y' && setdivision && typeof setdivision === 'string') {
			setdivision = setdivision.split('.');
			if (setdivision.length >= 3) {
				msgHeaders.setName = setdivision[0];
				msgHeaders.setArea = setdivision[1];
				msgHeaders.setID = setdivision.slice(2).join('.');

				msgHeaders.moduleName += '.' + msgHeaders.setName + msgHeaders.setArea + msgHeaders.setID;
			}
		}

		container = tafConfig.get('taf.application.container');

		if (tafConfig.get('taf.application.isdocker', '').toLowerCase() === 'y' && container && typeof container === 'string') {
			msgHeaders.sContainer = container;
		}
	} else {
		msgHeaders = {
			moduleName : 'NO_MODULE_NAME',
			ip : '127.0.0.1',
			iPropertyVer : 2
		};
	}
};

exports.init = init;

exports.create = function(obj, policies) {
	var headers = {}, key;

	if (!msgHeaders) {
		init();
	}

	Object.getOwnPropertyNames(msgHeaders).forEach(function(name) {
		headers[name] = msgHeaders[name];
	});

	if (typeof obj === 'string') {
		headers.propertyName = obj;
	} else {
		Object.getOwnPropertyNames(obj).forEach(function(name) {
			headers[name] = obj[name];
		});
	}

	headers.propertyName = headers.propertyName || 'NO_PROPERTY_NAME';

	key = Object.getOwnPropertyNames(headers).map(function(name) {
		return headers[name];
	}).join('.');
	
	data[key] = data[key] || new ReportObj(headers, policies);

	return data[key];
};

exports.POLICY = require('../policy');
