// @tencent/taf-monitor/property/PropertyFProxy.js
// **********************************************************************
// Parsed By TAFParser(3.2.2.11), Generated By Jce2Node(20180425)
// TAFParser Maintained By <WSRD> and Jce2Node Maintained By <superzheng>
// Generated from "PropertyF.jce" by Client Mode
// **********************************************************************

"use strict";

var assert    = require("assert");
var TafStream = require("@tencent/taf-stream");
var TafError  = require("@tencent/taf-rpc").error;

var taf = taf || {};
module.exports.taf = taf;

taf.PropertyFProxy = function () {
    this._name   = undefined;
    this._worker = undefined;
};

taf.PropertyFProxy.prototype.setTimeout = function (iTimeout) {
    this._worker.timeout = iTimeout;
};

taf.PropertyFProxy.prototype.getTimeout = function () {
    return this._worker.timeout;
};


taf.StatPropMsgHead = function() {
    this.moduleName = "";
    this.ip = "";
    this.propertyName = "";
    this.setName = "";
    this.setArea = "";
    this.setID = "";
    this.sContainer = "";
    this.iPropertyVer = 1;
    this._classname = "taf.StatPropMsgHead";
};
taf.StatPropMsgHead._classname = "taf.StatPropMsgHead";
taf.StatPropMsgHead._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.StatPropMsgHead._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.StatPropMsgHead._readFrom = function (is) {
    var tmp = new taf.StatPropMsgHead();
    tmp.moduleName = is.readString(0, true, "");
    tmp.ip = is.readString(1, true, "");
    tmp.propertyName = is.readString(2, true, "");
    tmp.setName = is.readString(3, false, "");
    tmp.setArea = is.readString(4, false, "");
    tmp.setID = is.readString(5, false, "");
    tmp.sContainer = is.readString(6, false, "");
    tmp.iPropertyVer = is.readInt32(7, false, 1);
    return tmp;
};
taf.StatPropMsgHead.prototype._writeTo = function (os) {
    os.writeString(0, this.moduleName);
    os.writeString(1, this.ip);
    os.writeString(2, this.propertyName);
    os.writeString(3, this.setName);
    os.writeString(4, this.setArea);
    os.writeString(5, this.setID);
    os.writeString(6, this.sContainer);
    os.writeInt32(7, this.iPropertyVer);
};
taf.StatPropMsgHead.prototype._equal = function (anItem) {
    return this.moduleName === anItem.moduleName && 
        this.ip === anItem.ip && 
        this.propertyName === anItem.propertyName && 
        this.setName === anItem.setName && 
        this.setArea === anItem.setArea && 
        this.setID === anItem.setID;
};
taf.StatPropMsgHead.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.StatPropMsgHead.prototype.toObject = function() { 
    return {
        "moduleName" : this.moduleName,
        "ip" : this.ip,
        "propertyName" : this.propertyName,
        "setName" : this.setName,
        "setArea" : this.setArea,
        "setID" : this.setID,
        "sContainer" : this.sContainer,
        "iPropertyVer" : this.iPropertyVer
    };
};
taf.StatPropMsgHead.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("moduleName") && (this.moduleName = json.moduleName);
    json.hasOwnProperty("ip") && (this.ip = json.ip);
    json.hasOwnProperty("propertyName") && (this.propertyName = json.propertyName);
    json.hasOwnProperty("setName") && (this.setName = json.setName);
    json.hasOwnProperty("setArea") && (this.setArea = json.setArea);
    json.hasOwnProperty("setID") && (this.setID = json.setID);
    json.hasOwnProperty("sContainer") && (this.sContainer = json.sContainer);
    json.hasOwnProperty("iPropertyVer") && (this.iPropertyVer = json.iPropertyVer);
};
taf.StatPropMsgHead.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.StatPropMsgHead.new = function () {
    return new taf.StatPropMsgHead();
};
taf.StatPropMsgHead.create = function (is) {
    return taf.StatPropMsgHead._readFrom(is);
};

taf.StatPropInfo = function() {
    this.policy = "";
    this.value = "";
    this._classname = "taf.StatPropInfo";
};
taf.StatPropInfo._classname = "taf.StatPropInfo";
taf.StatPropInfo._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.StatPropInfo._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.StatPropInfo._readFrom = function (is) {
    var tmp = new taf.StatPropInfo();
    tmp.policy = is.readString(0, true, "");
    tmp.value = is.readString(1, true, "");
    return tmp;
};
taf.StatPropInfo.prototype._writeTo = function (os) {
    os.writeString(0, this.policy);
    os.writeString(1, this.value);
};
taf.StatPropInfo.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.StatPropInfo.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.StatPropInfo.prototype.toObject = function() { 
    return {
        "policy" : this.policy,
        "value" : this.value
    };
};
taf.StatPropInfo.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("policy") && (this.policy = json.policy);
    json.hasOwnProperty("value") && (this.value = json.value);
};
taf.StatPropInfo.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.StatPropInfo.new = function () {
    return new taf.StatPropInfo();
};
taf.StatPropInfo.create = function (is) {
    return taf.StatPropInfo._readFrom(is);
};

taf.StatPropMsgBody = function() {
    this.vInfo = new TafStream.List(taf.StatPropInfo);
    this._classname = "taf.StatPropMsgBody";
};
taf.StatPropMsgBody._classname = "taf.StatPropMsgBody";
taf.StatPropMsgBody._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.StatPropMsgBody._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.StatPropMsgBody._readFrom = function (is) {
    var tmp = new taf.StatPropMsgBody();
    tmp.vInfo = is.readList(0, true, TafStream.List(taf.StatPropInfo));
    return tmp;
};
taf.StatPropMsgBody.prototype._writeTo = function (os) {
    os.writeList(0, this.vInfo);
};
taf.StatPropMsgBody.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.StatPropMsgBody.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.StatPropMsgBody.prototype.toObject = function() { 
    return {
        "vInfo" : this.vInfo.toObject()
    };
};
taf.StatPropMsgBody.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("vInfo") && (this.vInfo.readFromObject(json.vInfo));
};
taf.StatPropMsgBody.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.StatPropMsgBody.new = function () {
    return new taf.StatPropMsgBody();
};
taf.StatPropMsgBody.create = function (is) {
    return taf.StatPropMsgBody._readFrom(is);
};


var __taf_PropertyF$reportPropMsg$EN = function (statmsg) {
    var os = new TafStream.JceOutputStream();
    os.writeMap(1, statmsg);
    return os.getBinBuffer();
};

var __taf_PropertyF$reportPropMsg$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0)
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_PropertyF$reportPropMsg$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.PropertyFProxy.prototype.reportPropMsg = function (statmsg) {
    return this._worker.taf_invoke("reportPropMsg", __taf_PropertyF$reportPropMsg$EN(statmsg), arguments[arguments.length - 1]).then(__taf_PropertyF$reportPropMsg$DE, __taf_PropertyF$reportPropMsg$ER);
};

var __taf_PropertyF$reportTranPropMsg$EN = function (statmsg) {
    var os = new TafStream.JceOutputStream();
    os.writeMap(1, statmsg);
    return os.getBinBuffer();
};

var __taf_PropertyF$reportTranPropMsg$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0)
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_PropertyF$reportTranPropMsg$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.PropertyFProxy.prototype.reportTranPropMsg = function (statmsg) {
    return this._worker.taf_invoke("reportTranPropMsg", __taf_PropertyF$reportTranPropMsg$EN(statmsg), arguments[arguments.length - 1]).then(__taf_PropertyF$reportTranPropMsg$DE, __taf_PropertyF$reportTranPropMsg$ER);
};
