// @tencent/winston-taf/transport/taf-base.js
'use strict';

var path = require('path'),
	fs = require('fs'),
	assert = require('assert'),
	util = require('util');

var mkdirp = require('@tencent/super-mkdirp');

var winston = require('winston'),
	common = require('winston/lib/winston/common');

var Formatter = require('../lib/formatter');

var defaultMaxListeners = require('events').EventEmitter.defaultMaxListeners || 10;

var instances = {};

var TafBase = function (options) {
	assert(options.filename, 'Cannot log to file without filename');

	if (instances[options.filename]) {
		instances[options.filename].ref += 1;
		instances[options.filename].obj.setMaxListeners(instances[options.filename].ref * defaultMaxListeners);
		return instances[options.filename].obj;
	}

	winston.Transport.call(this, options);

	this.filename = options.filename;
	this.options = {
		interval : options.interval || 5 * 1000, //5s
		bufferSize : options.bufferSize || 10000,
		formatter : options.formatter || Formatter.Detail(),
		prefix : options.prefix
	};

	this._opening = false;
	this._dirchecked = false;
	this._buffer = [];
	this._bufferIndex = 0;

	this._orgname = this.filename;
	this._dirname = path.dirname(this.filename);
	this._extname = path.extname(this.filename);
	this._basename = path.basename(this.filename, this._extname);

	instances[this._orgname] = {
		ref : 1,
		obj : this
	};
};

util.inherits(TafBase, winston.Transport);

TafBase.prototype.name = 'tafBase';

TafBase.prototype.log = function (level, msg, meta, callback) {
	var output = common.log({
		level : level,
		message : msg,
		meta : meta,
		formatter : this.options.formatter,
		prefix : this.options.prefix
	}) + '\n';

	if (this._getFileStream()) {
		this._write(output, callback);
	} else {
		this._buffer[this._bufferIndex++] = {
			output : output,
			callback : callback
		};
		if (this._bufferIndex > this.options.bufferSize) {
			this._bufferIndex = 0;
		}
	}
};

TafBase.prototype._write = function(data, callback) {
	var ths = this;

	this._stream.write(data);

	if (!this._timerid) {
		this._timerid = setTimeout(function() {
			ths._closeStream();

			ths._timerid = undefined;
		}, this.options.interval);
	}

	callback(null, true);
};

TafBase.prototype._checkfile = function(cb) {
	this.emit('checkfile');
	
	fs.stat(this.filename, function(err, stats) {
		if (err) {
			if (err.code !== 'ENOENT') {
				cb(err);
				return;
			}
		}

		cb();
	});
};

TafBase.prototype._checkdir = function(cb) {
	var ths = this;

	if (this._dirchecked) {
		cb();
	} else {
		mkdirp(this._dirname, function(err) {
			ths._dirchecked = !err;
			cb(err);
		});
	}
};

TafBase.prototype._getFileStream = function() {
	var ths = this;

	if (this._stream) {
		return true;
	}

	if (this._opening) {
		return false;
	}

	this._opening = true;

	this._checkdir(function(err) {
		if (err) {
			ths.emit('error', err);
			return;
		}

		ths._checkfile(function(err) {
			var i = 0;

			if (err) {
				ths.emit('error', err);
				return;
			}

			setImmediate(function() {
				ths._stream = fs.createWriteStream(ths.filename, {
					flags : 'a'
				});
				ths._stream.on('error', function(err) {
					ths.emit('error', err);
				});

				for (i = ths._bufferIndex; i < ths._buffer.length; i += 1) {
					ths._write(ths._buffer[i].output, ths._buffer[i].callback);
				}
				for (i = 0; i < ths._bufferIndex; i += 1) {
					ths._write(ths._buffer[i].output, ths._buffer[i].callback);
				}
				ths._bufferIndex = 0;
				ths._buffer = [];
				
				ths._opening = false;
			});
		});
	});

	return false;
};

TafBase.prototype._closeStream = function() {
	if (this._stream) {
		this._stream.removeAllListeners('error');
		
		//fixme? maybe end with error?
		this._stream.end(function(_stream) {
			return function() {
				_stream.destroy();
				_stream = undefined;
			}
		}(this._stream));

		this._stream = undefined;
	}
};

TafBase.prototype.close = function() {
	if (this._timerid) {
		clearTimeout(this._timerid);
		this._timerid = undefined;
	}

	this._buffer = [];
	this._bufferIndex = 0;

	this._closeStream();

	if (instances[this._orgname]) {
		if (instances[this._orgname].ref > 1) {
			instances[this._orgname].ref -= 1;
			instances[this._orgname].obj.setMaxListeners(instances[this._orgname].ref * defaultMaxListeners);
		} else {
			delete instances[this._orgname];
		}
	}
};

module.exports = winston.transports.TafBase = TafBase;
