// @tencent/taf-monitor/stat/StatFProxy.js
// **********************************************************************
// Parsed By TAFParser(3.2.2.11), Generated By Jce2Node(20180425)
// TAFParser Maintained By <WSRD> and Jce2Node Maintained By <superzheng>
// Generated from "StatF.jce" by Client Mode
// **********************************************************************

"use strict";

var assert    = require("assert");
var TafStream = require("@tencent/taf-stream");
var TafError  = require("@tencent/taf-rpc").error;

var taf = taf || {};
module.exports.taf = taf;

taf.StatFProxy = function () {
    this._name   = undefined;
    this._worker = undefined;
};

taf.StatFProxy.prototype.setTimeout = function (iTimeout) {
    this._worker.timeout = iTimeout;
};

taf.StatFProxy.prototype.getTimeout = function () {
    return this._worker.timeout;
};


taf.StatMicMsgHead = function() {
    this.masterName = "";
    this.slaveName = "";
    this.interfaceName = "";
    this.masterIp = "";
    this.slaveIp = "";
    this.slavePort = 0;
    this.returnValue = 0;
    this.slaveSetName = "";
    this.slaveSetArea = "";
    this.slaveSetID = "";
    this.tafVersion = "";
    this.sMasterSetInfo = "";
    this.sSlaveContainer = "";
    this.sMasterContainer = "";
    this.iStatVer = 1;
    this._classname = "taf.StatMicMsgHead";
};
taf.StatMicMsgHead._classname = "taf.StatMicMsgHead";
taf.StatMicMsgHead._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.StatMicMsgHead._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.StatMicMsgHead._readFrom = function (is) {
    var tmp = new taf.StatMicMsgHead();
    tmp.masterName = is.readString(0, true, "");
    tmp.slaveName = is.readString(1, true, "");
    tmp.interfaceName = is.readString(2, true, "");
    tmp.masterIp = is.readString(3, true, "");
    tmp.slaveIp = is.readString(4, true, "");
    tmp.slavePort = is.readInt32(5, true, 0);
    tmp.returnValue = is.readInt32(6, true, 0);
    tmp.slaveSetName = is.readString(7, false, "");
    tmp.slaveSetArea = is.readString(8, false, "");
    tmp.slaveSetID = is.readString(9, false, "");
    tmp.tafVersion = is.readString(10, false, "");
    tmp.sMasterSetInfo = is.readString(11, false, "");
    tmp.sSlaveContainer = is.readString(12, false, "");
    tmp.sMasterContainer = is.readString(13, false, "");
    tmp.iStatVer = is.readInt32(14, false, 1);
    return tmp;
};
taf.StatMicMsgHead.prototype._writeTo = function (os) {
    os.writeString(0, this.masterName);
    os.writeString(1, this.slaveName);
    os.writeString(2, this.interfaceName);
    os.writeString(3, this.masterIp);
    os.writeString(4, this.slaveIp);
    os.writeInt32(5, this.slavePort);
    os.writeInt32(6, this.returnValue);
    os.writeString(7, this.slaveSetName);
    os.writeString(8, this.slaveSetArea);
    os.writeString(9, this.slaveSetID);
    os.writeString(10, this.tafVersion);
    os.writeString(11, this.sMasterSetInfo);
    os.writeString(12, this.sSlaveContainer);
    os.writeString(13, this.sMasterContainer);
    os.writeInt32(14, this.iStatVer);
};
taf.StatMicMsgHead.prototype._equal = function (anItem) {
    return this.masterName === anItem.masterName && 
        this.slaveName === anItem.slaveName && 
        this.interfaceName === anItem.interfaceName && 
        this.masterIp === anItem.masterIp && 
        this.slaveIp === anItem.slaveIp && 
        this.slavePort === anItem.slavePort && 
        this.returnValue === anItem.returnValue && 
        this.slaveSetName === anItem.slaveSetName && 
        this.slaveSetArea === anItem.slaveSetArea && 
        this.slaveSetID === anItem.slaveSetID && 
        this.tafVersion === anItem.tafVersion && 
        this.sMasterSetInfo === anItem.sMasterSetInfo && 
        this.sSlaveContainer === anItem.sSlaveContainer && 
        this.sMasterContainer === anItem.sMasterContainer && 
        this.iStatVer === anItem.iStatVer;
};
taf.StatMicMsgHead.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.StatMicMsgHead.prototype.toObject = function() { 
    return {
        "masterName" : this.masterName,
        "slaveName" : this.slaveName,
        "interfaceName" : this.interfaceName,
        "masterIp" : this.masterIp,
        "slaveIp" : this.slaveIp,
        "slavePort" : this.slavePort,
        "returnValue" : this.returnValue,
        "slaveSetName" : this.slaveSetName,
        "slaveSetArea" : this.slaveSetArea,
        "slaveSetID" : this.slaveSetID,
        "tafVersion" : this.tafVersion,
        "sMasterSetInfo" : this.sMasterSetInfo,
        "sSlaveContainer" : this.sSlaveContainer,
        "sMasterContainer" : this.sMasterContainer,
        "iStatVer" : this.iStatVer
    };
};
taf.StatMicMsgHead.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("masterName") && (this.masterName = json.masterName);
    json.hasOwnProperty("slaveName") && (this.slaveName = json.slaveName);
    json.hasOwnProperty("interfaceName") && (this.interfaceName = json.interfaceName);
    json.hasOwnProperty("masterIp") && (this.masterIp = json.masterIp);
    json.hasOwnProperty("slaveIp") && (this.slaveIp = json.slaveIp);
    json.hasOwnProperty("slavePort") && (this.slavePort = json.slavePort);
    json.hasOwnProperty("returnValue") && (this.returnValue = json.returnValue);
    json.hasOwnProperty("slaveSetName") && (this.slaveSetName = json.slaveSetName);
    json.hasOwnProperty("slaveSetArea") && (this.slaveSetArea = json.slaveSetArea);
    json.hasOwnProperty("slaveSetID") && (this.slaveSetID = json.slaveSetID);
    json.hasOwnProperty("tafVersion") && (this.tafVersion = json.tafVersion);
    json.hasOwnProperty("sMasterSetInfo") && (this.sMasterSetInfo = json.sMasterSetInfo);
    json.hasOwnProperty("sSlaveContainer") && (this.sSlaveContainer = json.sSlaveContainer);
    json.hasOwnProperty("sMasterContainer") && (this.sMasterContainer = json.sMasterContainer);
    json.hasOwnProperty("iStatVer") && (this.iStatVer = json.iStatVer);
};
taf.StatMicMsgHead.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.StatMicMsgHead.new = function () {
    return new taf.StatMicMsgHead();
};
taf.StatMicMsgHead.create = function (is) {
    return taf.StatMicMsgHead._readFrom(is);
};

taf.StatMicMsgBody = function() {
    this.count = 0;
    this.timeoutCount = 0;
    this.execCount = 0;
    this.intervalCount = new TafStream.Map(TafStream.Int32, TafStream.Int32);
    this.totalRspTime = 0;
    this.maxRspTime = 0;
    this.minRspTime = 0;
    this.retValue = new TafStream.Map(TafStream.Int64, TafStream.Int32);
    this.weightValue = 0;
    this.weightCount = 0;
    this._classname = "taf.StatMicMsgBody";
};
taf.StatMicMsgBody._classname = "taf.StatMicMsgBody";
taf.StatMicMsgBody._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.StatMicMsgBody._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.StatMicMsgBody._readFrom = function (is) {
    var tmp = new taf.StatMicMsgBody();
    tmp.count = is.readInt32(0, true, 0);
    tmp.timeoutCount = is.readInt32(1, true, 0);
    tmp.execCount = is.readInt32(2, true, 0);
    tmp.intervalCount = is.readMap(3, true, TafStream.Map(TafStream.Int32, TafStream.Int32));
    tmp.totalRspTime = is.readInt64(4, true, 0);
    tmp.maxRspTime = is.readInt32(5, true, 0);
    tmp.minRspTime = is.readInt32(6, true, 0);
    tmp.retValue = is.readMap(7, false, TafStream.Map(TafStream.Int64, TafStream.Int32));
    tmp.weightValue = is.readInt32(8, false, 0);
    tmp.weightCount = is.readInt32(9, false, 0);
    return tmp;
};
taf.StatMicMsgBody.prototype._writeTo = function (os) {
    os.writeInt32(0, this.count);
    os.writeInt32(1, this.timeoutCount);
    os.writeInt32(2, this.execCount);
    os.writeMap(3, this.intervalCount);
    os.writeInt64(4, this.totalRspTime);
    os.writeInt32(5, this.maxRspTime);
    os.writeInt32(6, this.minRspTime);
    os.writeMap(7, this.retValue);
    os.writeInt32(8, this.weightValue);
    os.writeInt32(9, this.weightCount);
};
taf.StatMicMsgBody.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.StatMicMsgBody.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.StatMicMsgBody.prototype.toObject = function() { 
    return {
        "count" : this.count,
        "timeoutCount" : this.timeoutCount,
        "execCount" : this.execCount,
        "intervalCount" : this.intervalCount.toObject(),
        "totalRspTime" : this.totalRspTime,
        "maxRspTime" : this.maxRspTime,
        "minRspTime" : this.minRspTime,
        "retValue" : this.retValue.toObject(),
        "weightValue" : this.weightValue,
        "weightCount" : this.weightCount
    };
};
taf.StatMicMsgBody.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("count") && (this.count = json.count);
    json.hasOwnProperty("timeoutCount") && (this.timeoutCount = json.timeoutCount);
    json.hasOwnProperty("execCount") && (this.execCount = json.execCount);
    json.hasOwnProperty("intervalCount") && (this.intervalCount.readFromObject(json.intervalCount));
    json.hasOwnProperty("totalRspTime") && (this.totalRspTime = json.totalRspTime);
    json.hasOwnProperty("maxRspTime") && (this.maxRspTime = json.maxRspTime);
    json.hasOwnProperty("minRspTime") && (this.minRspTime = json.minRspTime);
    json.hasOwnProperty("retValue") && (this.retValue.readFromObject(json.retValue));
    json.hasOwnProperty("weightValue") && (this.weightValue = json.weightValue);
    json.hasOwnProperty("weightCount") && (this.weightCount = json.weightCount);
};
taf.StatMicMsgBody.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.StatMicMsgBody.new = function () {
    return new taf.StatMicMsgBody();
};
taf.StatMicMsgBody.create = function (is) {
    return taf.StatMicMsgBody._readFrom(is);
};

taf.StatSampleMsg = function() {
    this.unid = "";
    this.masterName = "";
    this.slaveName = "";
    this.interfaceName = "";
    this.masterIp = "";
    this.slaveIp = "";
    this.depth = 0;
    this.width = 0;
    this.parentWidth = 0;
    this._classname = "taf.StatSampleMsg";
};
taf.StatSampleMsg._classname = "taf.StatSampleMsg";
taf.StatSampleMsg._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.StatSampleMsg._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.StatSampleMsg._readFrom = function (is) {
    var tmp = new taf.StatSampleMsg();
    tmp.unid = is.readString(0, true, "");
    tmp.masterName = is.readString(1, true, "");
    tmp.slaveName = is.readString(2, true, "");
    tmp.interfaceName = is.readString(3, true, "");
    tmp.masterIp = is.readString(4, true, "");
    tmp.slaveIp = is.readString(5, true, "");
    tmp.depth = is.readInt32(6, true, 0);
    tmp.width = is.readInt32(7, true, 0);
    tmp.parentWidth = is.readInt32(8, true, 0);
    return tmp;
};
taf.StatSampleMsg.prototype._writeTo = function (os) {
    os.writeString(0, this.unid);
    os.writeString(1, this.masterName);
    os.writeString(2, this.slaveName);
    os.writeString(3, this.interfaceName);
    os.writeString(4, this.masterIp);
    os.writeString(5, this.slaveIp);
    os.writeInt32(6, this.depth);
    os.writeInt32(7, this.width);
    os.writeInt32(8, this.parentWidth);
};
taf.StatSampleMsg.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.StatSampleMsg.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.StatSampleMsg.prototype.toObject = function() { 
    return {
        "unid" : this.unid,
        "masterName" : this.masterName,
        "slaveName" : this.slaveName,
        "interfaceName" : this.interfaceName,
        "masterIp" : this.masterIp,
        "slaveIp" : this.slaveIp,
        "depth" : this.depth,
        "width" : this.width,
        "parentWidth" : this.parentWidth
    };
};
taf.StatSampleMsg.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("unid") && (this.unid = json.unid);
    json.hasOwnProperty("masterName") && (this.masterName = json.masterName);
    json.hasOwnProperty("slaveName") && (this.slaveName = json.slaveName);
    json.hasOwnProperty("interfaceName") && (this.interfaceName = json.interfaceName);
    json.hasOwnProperty("masterIp") && (this.masterIp = json.masterIp);
    json.hasOwnProperty("slaveIp") && (this.slaveIp = json.slaveIp);
    json.hasOwnProperty("depth") && (this.depth = json.depth);
    json.hasOwnProperty("width") && (this.width = json.width);
    json.hasOwnProperty("parentWidth") && (this.parentWidth = json.parentWidth);
};
taf.StatSampleMsg.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.StatSampleMsg.new = function () {
    return new taf.StatSampleMsg();
};
taf.StatSampleMsg.create = function (is) {
    return taf.StatSampleMsg._readFrom(is);
};

taf.ProxyInfo = function() {
    this.bFromClient = true;
    this._classname = "taf.ProxyInfo";
};
taf.ProxyInfo._classname = "taf.ProxyInfo";
taf.ProxyInfo._write = function (os, tag, value) { os.writeStruct(tag, value); };
taf.ProxyInfo._read  = function (is, tag, def) { return is.readStruct(tag, true, def); };
taf.ProxyInfo._readFrom = function (is) {
    var tmp = new taf.ProxyInfo();
    tmp.bFromClient = is.readBoolean(0, true, true);
    return tmp;
};
taf.ProxyInfo.prototype._writeTo = function (os) {
    os.writeBoolean(0, this.bFromClient);
};
taf.ProxyInfo.prototype._equal = function () {
    assert.fail("this structure not define key operation");
};
taf.ProxyInfo.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
};
taf.ProxyInfo.prototype.toObject = function() { 
    return {
        "bFromClient" : this.bFromClient
    };
};
taf.ProxyInfo.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("bFromClient") && (this.bFromClient = json.bFromClient);
};
taf.ProxyInfo.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
};
taf.ProxyInfo.new = function () {
    return new taf.ProxyInfo();
};
taf.ProxyInfo.create = function (is) {
    return taf.ProxyInfo._readFrom(is);
};


var __taf_StatF$reportMicMsg$EN = function (msg, bFromClient) {
    var os = new TafStream.JceOutputStream();
    os.writeMap(1, msg);
    os.writeBoolean(2, bFromClient);
    return os.getBinBuffer();
};

var __taf_StatF$reportMicMsg$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0)
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_StatF$reportMicMsg$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.StatFProxy.prototype.reportMicMsg = function (msg, bFromClient) {
    return this._worker.taf_invoke("reportMicMsg", __taf_StatF$reportMicMsg$EN(msg, bFromClient), arguments[arguments.length - 1]).then(__taf_StatF$reportMicMsg$DE, __taf_StatF$reportMicMsg$ER);
};

var __taf_StatF$reportProxyMicMsg$EN = function (msg, proxyInfo) {
    var os = new TafStream.JceOutputStream();
    os.writeMap(1, msg);
    os.writeStruct(2, proxyInfo);
    return os.getBinBuffer();
};

var __taf_StatF$reportProxyMicMsg$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0)
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_StatF$reportProxyMicMsg$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.StatFProxy.prototype.reportProxyMicMsg = function (msg, proxyInfo) {
    return this._worker.taf_invoke("reportProxyMicMsg", __taf_StatF$reportProxyMicMsg$EN(msg, proxyInfo), arguments[arguments.length - 1]).then(__taf_StatF$reportProxyMicMsg$DE, __taf_StatF$reportProxyMicMsg$ER);
};

var __taf_StatF$reportSampleMsg$EN = function (msg) {
    var os = new TafStream.JceOutputStream();
    os.writeList(1, msg);
    return os.getBinBuffer();
};

var __taf_StatF$reportSampleMsg$DE = function (data) {
    try {
        var is = new TafStream.JceInputStream(data.response.sBuffer);
        return {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "return" : is.readInt32(0, true, 0)
            }
        };
    } catch (e) {
        throw {
            "request" : data.request,
            "response" : {
                "costtime" : data.request.costtime,
                "error" : {
                    "code" : TafError.CLIENT.DECODE_ERROR,
                    "message" : e.message
                }
            }
        };
    }
};

var __taf_StatF$reportSampleMsg$ER = function (data) {
    throw {
        "request" : data.request,
        "response" : {
            "costtime" : data.request.costtime,
            "error" : data.error
        }
    }
};

taf.StatFProxy.prototype.reportSampleMsg = function (msg) {
    return this._worker.taf_invoke("reportSampleMsg", __taf_StatF$reportSampleMsg$EN(msg), arguments[arguments.length - 1]).then(__taf_StatF$reportSampleMsg$DE, __taf_StatF$reportSampleMsg$ER);
};
