// @tencent/taf-rpc/examples/taf-rpc/demo.1/server.node.3/NodeTafImp.js
// **********************************************************************
// This file was generated by a TAF parser!
// TAF version 3.1.1.2 by WSRD Tencent.
// Generated from `publish.jce'
// **********************************************************************

var TRom = require('./NodeTaf.js').TRom;
module.exports.TRom = TRom;

TRom.NodeTafImp.prototype.initialize = function ( ) {
    //TODO::
    //this.aPrx = this.application.getCommunicator().stringToProxy(TRom, "TRom.APproxy.TestObj@127.0.0.1");
    console.log("NodeTafImp::initialize");
}

TRom.NodeTafImp.prototype.test = function (current) {
    //TODO::

}

TRom.NodeTafImp.prototype.getall = function (current, stUser, stResult) {
    //TODO::
    console.log("stUser.name::", stUser.name);

    stResult.iLevel = 11000;
    current.sendResponse(10000, stResult);
}

TRom.NodeTafImp.prototype.getUsrName = function (current, sUsrName, sValue1, sValue2) {
    //TODO::
    console.log("sUsrName::", sUsrName);
    sValue1 = "OUT_" + sUsrName + "_1";
    sValue2 = "OUT_" + sUsrName + "_2";

    current.sendResponse(100002, sValue1, sValue2);
}
